packages/react-dom/src/events/ReactDOMEventListener.js

```diff
@@ -155,12 +155,37 @@ export function dispatchEvent(
     return;
   }
 
-  // TODO: replaying capture phase events is currently broken
-  // because we used to do it during top-level native bubble handlers
-  // but now we use different bubble and capture handlers.
-  // In eager mode, we attach capture listeners early, so we need
-  // to filter them out until we fix the logic to handle them correctly.
-  const allowReplay = (eventSystemFlags & IS_CAPTURE_PHASE) === 0;
+  const allowReplay =
+    enableCapturePhaseSelectiveHydrationWithoutDiscreteEventReplay ||
+    (eventSystemFlags & IS_CAPTURE_PHASE) === 0;
+
+  let blockedOn = findInstanceBlockingEvent(
+    domEventName,
+    eventSystemFlags,
+    targetContainer,
+    nativeEvent,
+  );
+
+  // We can dispatch the event now
```

Comment:
A bit of meta feedback. This function is getting pretty difficult to read because the behavior with the flag and without is so different. Some conditions are already repeated.

How would you feel about completely forking the body of this function into two paths?

```js
if (enableCapturePhaseSelectiveHydrationWithoutDiscreteEventReplay) {
 // this code
} else {
  // that code
}
```

The duplication between them is okay but I think this would help understand what each code path is doing. If some parts don't execute at all (e.g. discrete stuff inside the flag) then we can just delete that part inside the `if` and make it simpler to follow.